### PR TITLE
Support latest service-catalog in OS 3.7

### PR DESCRIPTION
**Pull request by djzager**
Update the ansible service broker setup to wait for the service-catalog
to be available. This prevents a race condition when we are working with
coalmine and the service-catalog (giving it time to redeploy).

### CODE REVIEW

**Comment by jmontleon** [comment]
```
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "/home/jmontleo/bin/oc patch -n service-catalog deployment/apiserver -p '{\"spec\":{\"template\": {\"spec\": {\"containers\": [{\"name\":\"apiserver\",\"image\":\"quay.io/kubernetes-service-catalog/apiserver:canary\", \"command\": [\"/opt/services/apiserver\"] }]}}}}'", "delta": "0:00:00.220858", "end": "2017-09-06 09:11:43.732962", "failed": true, "rc": 1, "start": "2017-09-06 09:11:43.512104", "stderr": "Error from server (NotFound): namespaces \"service-catalog\" not found", "stderr_lines": ["Error from server (NotFound): namespaces \"service-catalog\" not found"], "stdout": "", "stdout_lines": []}
```

With these settings:
```
coalmine: true
coalmine_apiserver: quay.io/kubernetes-service-catalog/apiserver:canary
coalmine_controller: quay.io/kubernetes-service-catalog/controller-manager:canary
origin_image_tag: latest
openshift_client_version: '3.7'
```

Pods like to be under kube-service-catalog rather than service-catalog
```
kube-service-catalog                apiserver-2314068821-q6lbz           2/2       Running     0          1m
kube-service-catalog                controller-manager-340623315-zfvd1   1/1       Running     1          1m
```

File: ansible/roles/env_hacks/tasks/main.yml [hunk 1]
```diff
@@ -25,21 +25,36 @@
...
+            shell: "{{ oc_cmd }} patch -n service-catalog deployment/controller-manager -p '{\"spec\":{\"template\": {\"spec\": {\"containers\": [{\"name\":\"controller-manager\",\"image\":\"{{ coalmine_controller }}\", \"command\": [\"/opt/services/controller-manager\"], \"args\": {{ arglist.stdout }} }]}}}}'"
+        when: openshift_client_version == '3.6'
+
+      - name: Update service-catalog (OpenShift 3.7)
+        block:
+        - name: Workaround canary permissions issue
+          shell: "{{ oc_cmd }} adm policy add-cluster-role-to-user cluster-admin system:anonymous"
+
+        - name: Use canary apiserver
+          shell: "{{ oc_cmd }} patch -n service-catalog deployment/apiserver -p '{\"spec\":{\"template\": {\"spec\": {\"containers\": [{\"name\":\"apiserver\",\"image\":\"{{ coalmine_apiserver }}\", \"command\": [\"/opt/services/apiserver\"] }]}}}}'"
```
Comment (jmontleon) [inline ansible/roles/env_hacks/tasks/main.yml]: -n kube-service-catalog worked for me. I'm not sure if this is a (very) recent change? I didn't have this issue the other day when we hit this, I don't believe.

File: ansible/roles/env_hacks/tasks/main.yml [hunk 2]
```diff
@@ -25,21 +25,36 @@
...
+
+        - name: Use canary controller-manager
+          shell: "{{ oc_cmd }} patch -n service-catalog deployment/controller-manager -p '{\"spec\":{\"template\": {\"spec\": {\"containers\": [{\"name\":\"controller-manager\",\"image\":\"{{ coalmine_controller }}\", \"command\": [\"/opt/services/controller-manager\"]}]}}}}'"
```
Comment (jmontleon) [inline ansible/roles/env_hacks/tasks/main.yml]: -n kube-service-katalog here as well.

File: ansible/roles/ansible_service_broker_setup/tasks/main.yml [hunk 3]
```diff
@@ -1,4 +1,13 @@
 ---
+  - name: Waiting 1 minutes for service-catalog
+    action: >-
+      shell "{{ oc_cmd }}" get deployments -n service-catalog apiserver
```
Comment (jmontleon) [inline ansible/roles/ansible_service_broker_setup/tasks/main.yml]: -n kube-service-catalog worked here for me.

File: ansible/roles/env_hacks/tasks/main.yml (hunk 1 above)
Comment (djzager) [reply ansible/roles/env_hacks/tasks/main.yml]: I don't remember it being in `kube-service-catalog`. I'll go another round with the latest and see.

File: ansible/roles/ansible_service_broker_setup/tasks/main.yml (hunk 3 above)
Comment (jmontleon) [reply ansible/roles/ansible_service_broker_setup/tasks/main.yml]: This patch in conjunction with these changes to the template gave me a successful run with coalmine and 3.7, though I did not see any APB's populated:
https://github.com/openshift/ansible-service-broker/pull/422

```
diff --git a/ansible/roles/ansible_service_broker_setup/tasks/main.yml b/ansible/roles/ansible_service_broker_setup/tasks/main.yml
index 335353d..2e712db 100644
--- a/ansible/roles/ansible_service_broker_setup/tasks/main.yml
+++ b/ansible/roles/ansible_service_broker_setup/tasks/main.yml
@@ -83,6 +83,14 @@
         -p SANDBOX_ROLE={{ broker_sandbox_role }}
     when: "not rcm"
 
+  - name: Update Broker Kind and Auth when using coalmine
+    set_fact:
+      cmd_process_asb_template: >-
+        {{ cmd_process_asb_template }}
+        -p BROKER_KIND="ServiceBroker"
+        -p BROKER_AUTH='{ "basic": { "secretRef": { "namespace": "ansible-service-broker", "name": "asb-auth-secret" } } }'
+    when: coalmine is defined and coalmine
+
   - name: Process template "{{ local_target_asb_template }}"
     shell: >
       {{ cmd_process_asb_template }}

```

**Comment by djzager** [comment]
Just to keep the comments up to date...It looks like `service-catalog` project has been renamed to `kube-service-catalog`. Catasb should be able to handle this and changes have been made in this PR to make it discover which project the `service-catalog` exists under and use that throughout setup.

**Comment by djzager** [comment]
Testing OpenShift 3.7 with latest service-catalog, config variables look like:

```
dockerhub_user_name: changeme
dockerhub_user_password: changeme
dockerhub_org_name: ansibleplaybookbundle

# Use jmontleo's template
asb_template_url: https://raw.githubusercontent.com/jmontleon/ansible-service-broker/template-for-3.7/templates/deploy-ansible-service-broker.template.yaml

# Use latest OpenShift Origin
origin_image_tag: latest
openshift_client_version: '3.7'

# Use coalmine to get latest service-catalog
coalmine: true
coalmine_apiserver: quay.io/kubernetes-service-catalog/apiserver:canary
coalmine_controller: quay.io/kubernetes-service-catalog/controller-manager:canary

# Override broker kind and broker auth section
broker_enable_basic_auth: true
broker_kind: ServiceBroker
broker_auth: '{ "basic": { "secretRef": { "namespace": "ansible-service-broker", "name": "asb-auth-secret" }}}'
```

File: ansible/roles/ansible_service_broker_setup/tasks/main.yml [hunk 4]
```diff
@@ -72,6 +81,8 @@
         -p REFRESH_INTERVAL={{ broker_bootstrap_refresh_interval }}
         -p TAG={{ apbtag | default('latest') }}
         -p SANDBOX_ROLE={{ broker_sandbox_role }}
+        {% if broker_kind is defined %}-p BROKER_KIND={{ broker_kind }}{% endif %}
```
Comment (jmontleon) [inline ansible/roles/ansible_service_broker_setup/tasks/main.yml]: Why not do this for the user if they set coalmine? It's bad enough to have to set 4 entries right now to enable http and I'm thinking either to set those for the user when they run coalmine, or at least cut it down to one ssl option that sets all 4. Having to flip 3 or 4 flags to do one thing is mean. 

```
+  - name: Update Broker Kind and Auth when using coalmine
+    set_fact:
+      cmd_process_asb_template: >-
+        {{ cmd_process_asb_template }}
+        -p BROKER_KIND="ServiceBroker"
+        -p BROKER_AUTH='{ "basic": { "secretRef": { "namespace": "ansible-service-broker", "name": "asb-auth-secret" } } }'
+    when: coalmine is defined and coalmine
```

**Comment by jmontleon** [comment]
One more change and I think we're good. Maybe address the comment above.

```
diff --git a/ansible/group_vars/all.yml b/ansible/group_vars/all.yml
index 718105b..ed780ce 100644
--- a/ansible/group_vars/all.yml
+++ b/ansible/group_vars/all.yml
@@ -111,6 +111,7 @@ broker_output_request: false
 broker_recovery: true
 broker_enable_basic_auth: true
 broker_insecure_mode: false
+broker_insecure_edge: None
 broker_endpoint_termination: Reencrypt
 broker_bootstrap_refresh_interval: 600s
 broker_sandbox_role: edit
diff --git a/ansible/roles/ansible_service_broker_setup/tasks/main.yml b/ansible/roles/ansible_service_broker_setup/tasks/main.yml
index c5125e9..f9b8e02 100644
--- a/ansible/roles/ansible_service_broker_setup/tasks/main.yml
+++ b/ansible/roles/ansible_service_broker_setup/tasks/main.yml
@@ -77,6 +77,7 @@
         -p REGISTRY_NAME={{ broker_registry_name }}
         -p ENABLE_BASIC_AUTH={{ broker_enable_basic_auth }}
         -p INSECURE={{ broker_insecure_mode|lower }}
+        -p INSECURE_EDGE={{ broker_insecure_edge }}
         -p TERMINATION={{ broker_endpoint_termination }}
         -p REFRESH_INTERVAL={{ broker_bootstrap_refresh_interval }}
         -p TAG={{ apbtag | default('latest') }}
```

My config for testing:
```
---
###############################
#Account                      #
###############################
dockerhub_user_name: changeme
dockerhub_user_password: changeme
dockerhub_org_name: ansibleplaybookbundle

###############################
#3.7                          #
###############################
origin_image_tag: latest
openshift_client_version: '3.7'

###############################
#Coalmine                     #
###############################
coalmine: true
coalmine_apiserver: quay.io/kubernetes-service-catalog/apiserver:canary
coalmine_controller: quay.io/kubernetes-service-catalog/controller-manager:canary
broker_kind: ServiceBroker
broker_auth: '{ "basic": { "secretRef": { "namespace": "ansible-service-broker", "name": "asb-auth-secret" }}}'

###############################
#http                         #
###############################
asb_scheme: http
broker_endpoint_termination: edge
broker_insecure_mode: true
broker_insecure_edge: Allow
```

**Comment by djzager** [comment]
The `my_vars.yml` used to get OS 3.7 cluster + coalmine service catalog running:

```
#################
# Docker login
#################
dockerhub_user_name: changeme
dockerhub_user_password: changeme
dockerhub_org_name: ansibleplaybookbundle

asb_template_url: https://raw.githubusercontent.com/jmontleon/ansible-service-broker/template-for-3.7/templates/deploy-ansible-service-broker.template.yaml

#################
# 3.7
#################
origin_image_tag: latest
openshift_client_version: '3.7'

#################
# Latest service-catalog
#################
coalmine: true

#################
# http
#################
asb_scheme: http
broker_endpoint_termination: edge
broker_insecure_mode: true
broker_insecure_edge: Allow
```

**Comment by djzager** [comment]
After simplifying on/off:

```
---

dockerhub_user_name: changeme
dockerhub_user_password: changeme
dockerhub_org_name: ansibleplaybookbundle

# Use jmontleon's template
asb_template_url: https://raw.githubusercontent.com/jmontleon/ansible-service-broker/template-for-3.7/templates/deploy-ansible-service-broker.template.yaml

#################
# 3.7
#################
origin_image_tag: latest
openshift_client_version: '3.7'

#################
# Latest service-catalog
#################
coalmine: true
broker_ssl: false # since latest doesn't support ssl at the moment
```

**Comment by djzager** [comment]
This looks good to me. The only thing I could even suggest would be to update `docs/auth.md` to mention that newer service catalogs look like: 

```
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: ServiceBroker
metadata:
  name: ansible-service-broker
spec:
  url: https://asb-1338-ansible-service-broker.172.17.0.1.nip.io
  authInfo:
    basic:
      secretRef:
        namespace: ansible-service-broker
        name: asb-auth-secret
```

**Comment by djzager** [comment]
[(0.5) Update BasicAuth config to work with latest Service Catalog](https://trello.com/c/pSaPdF1i/460-05-update-basicauth-config-to-work-with-latest-service-catalog)

**Comment by jmontleon** [comment]
ACK

**Pull request merged**